File: ALGORITMOS/java/QuickSort.java
//Questão 5
//Ordenação: Ordenar um array usando o método Quicksort

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class QuickSort {
  
    public static int[] quickSort(int[] arr) {
        if (arr.length <= 1) {
            return arr;
        }

        int pivot = arr[0];
        List<Integer> left = new ArrayList<>();
        List<Integer> right = new ArrayList<>();

        for (int i = 1; i < arr.length; i++) {
            if (arr[i] < pivot) {
                left.add(arr[i]);
            } else {
                right.add(arr[i]);
            }
        }

        int[] sortedLeft = quickSort(listToArray(left));
        int[] sortedRight = quickSort(listToArray(right));

        return concatArrays(sortedLeft, pivot, sortedRight);
    }

    private static int[] listToArray(List<Integer> list) {
        int[] result = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            result[i] = list.get(i);
        }
        return result;
    }

    private static int[] concatArrays(int[] left, int pivot, int[] right) {
        int[] result = new int[left.length + 1 + right.length];
        int index = 0;

        for (int val : left) {
            result[index++] = val;
        }

        result[index++] = pivot;

        for (int val : right) {
            result[index++] = val;
        }

        return result;
    }
}

File: ALGORITMOS/javascript/quickSort.js
//Questão 5
//Ordenar um array usando o método Quicksort

function quickSort(arr) {
    if (arr.length <= 1) {
      return arr;
    }
  
    const pivot = arr[0];
    const left = [];
    const right = [];
  
    for (let i = 1; i < arr.length; i++) {
      if (arr[i] < pivot) {
        left.push(arr[i]);
      } else {
        right.push(arr[i]);
      }
    }
    return [...quickSort(left), pivot, ...quickSort(right)];
  }

module.exports = quickSort;

let arrayDesordenado = [5, 3, 8, 1, 9, 4, 2, 7, 6];
console.log( "Array desordenado:", arrayDesordenado );
console.log( "Array ordenado:", quickSort(arrayDesordenado) );
